[v0.16.x] v0.16 release notes (#3582)
dashpay/dash: doc/release-notes.md

Dash Core version 0.16
======================

Release is now available from:

  <https://www.dash.org/downloads/#wallets>

This is a new major version release, bringing new features, various bugfixes

> and other improvements.

Reviewer: ```suggestion
and other improvements.

This version is the result of around 7 months of hard work consisting of more than 270 pull requests and totaling more than 900 commits.
```

Author: These are pretty vague metrics imo (not much better than counting deleted/added lines), would rather avoid that.

Reviewer: Okay, just a thought

> This release is mandatory for all nodes.

Reviewer 2: Doesn't version make more sense here?

```suggestion
This version is mandatory for all nodes.
```

Reviewer: Release is better to me

The "Pay To" field on "Send" and "PrivateSend" tabs accepts not only plain
Dash addresses but also Dash URIs now. The Dash address and the amount from
the URI are assigned to corresponding fields automatically.

Sporks
------
`SPORK_21_QUORUM_ALL_CONNECTED` was added to control Concentrated Recovery and
Masternode Probing activation. Sporks `SPORK_15_DETERMINISTIC_MNS_ENABLED`,
`SPORK_16_INSTANTSEND_AUTOLOCKS` and `SPORK_20_INSTANTSEND_LLMQ_BASED` which
were previously deprecated in v0.15 are completely removed now.

Build system
------------
The minimum version of the GCC compiler required to compile Dash Core is now 4.8.
The minimum version of Qt is now 5.5.1. Some packages in `depends/` as well as
`secp256k1` and `leveldb` subtrees were updated to newer versions.

RPC changes
-----------
There are a few changes in existing RPC interfaces in this release:
- `getpeerinfo` has new field `masternode` to indicate whether connection was
  due to masternode connection attempt
- `getprivatesendinfo` `denoms` field was replaced by `denoms_goal` and
  `denoms_hardcap`
- `listunspent` has new filter option `coinType` to be able to filter different
  types of coins (all, mixed etc.)

> - `quorum dkgstatus` shows `quorumConnections` for each LLMQ with detailed

Reviewer 2: ```suggestion
- `protx diff` returns more detailed information about new quorums
- `quorum dkgstatus` shows `quorumConnections` for each LLMQ with detailed
```

  information about each participating masternode
- `quorum sign` has an optional `quorumHash` argument to pick the exact quorum
- `socketevents` in `getnetworkinfo` rpc shows the socket events mode,
  either `epoll`, `poll` or `select`

There are also new RPC commands:
- `quorum selectquorum` returns the quorum that would/should sign a request

There are also new RPC commands backported from Bitcoin Core 0.16:
- `help-console` for more info about using console in Qt

> - `loadwallet` loads a wallet from a wallet file or directory

Reviewer: Should probably go into more detail about this?

of the network hashrate to signal update readiness as possible, this can result in
quite lengthy upgrades if one large non-upgraded entity stalls
all progress. Simply lowering thresholds even further can result in network
upgrades being too fast which can potentially cause some chaos. This version
implements BIP9-like dynamic activation thresholds which drop from some initial
level to a minimally acceptable one over time at an increasing rate. This provides
a safe non-blocking way of activating proposals.

This mechanism applies to the Block Reward Reallocation proposal mentioned above
for which the threshold will begin at an 80% level and decay down to 60% over the
course of 10 periods.

Concentrated Recovery
---------------------
In the current system, signature shares are propagated to all LLMQ members
until one of them has enough shares collected to recover the signature. Until
this recovered signature is propagated in the LLMQ, all members will keep
propagating shares and verifying each one. This causes significant load on the
LLMQ, resulting in decreased throughput, which will be avoided with the new system.

> The new system initially sends all shares to a single deterministically selected node,

Author: We were talking about the old system in the paragraph above. Using "this" feels weird here, no?

> The new system initially sends all shares to a single deterministically selected node,

Reviewer: which will be avoided with the new system. -> The new system initially felt weird to me, you ended the paragraph hat I think it probably makes sense to use this. 

Although this is pretty minor

Some potential attacks on PrivateSend assume that all inputs had been mixed
for the same number of rounds (up to 16). While this assumption alone is not
enough to break PrivateSend privacy, it could still provide some additional
information for other methods like cluster analysis and help to narrow results.

With Random Round Mixing, an input will be mixed to N rounds like prior. However,
at this point, SHA256(input, salt) will be calculated and roughly a half of inputs
will be mixed for one more round and this rule will be applied to newly created
inputs again and so on. This results in an exponential decay where if you mix a set
of inputs, half of those inputs will be mixed for N rounds, 1/4 will be mixed N+1,
1/8 will be mixed N+2, etc. Current implementation caps it at N+3 which results
in mixing an average of N+0.875 rounds and sounds like a reasonable compromise
given the privacy gains.

GUI changes
-----------
Many dialogs, tabs, icons, colors and various interface elements were reworked
to improve user experience and to make wallet look more consistent. The wallet
GUI is also much more flexible now. For regular users there is a new "Appearance"

> dialog and a corresponding tab in Options which allows them to pick a theme and

Reviewer: ```suggestion
dialog and a corresponding tab in the options which allows them to pick a theme and
```

A new algorithm for the creation of mixing denominations was implemented which
should reduce the number of the smallest inputs created and give users more
control on soft and hard caps. A much more accurate fee management algorithm was
also implemented which should fix issues for users who have custom fees
specified in wallet config or in times when network is under load.

There is a new "PrivateSend" tab in the GUI which allows spending fully
mixed coins only. The CoinControl feature was also improved to display coins
based on the tab it was opened in, to protect users from accidentally spending
mixed and non-mixed coins in the same transaction and to give better overview of
spendable mixed coins.

PrivateSend Random Round Mixing
-------------------------------
Some potential attacks on PrivateSend assume that all inputs had been mixed
for the same number of rounds (up to 16). While this assumption alone is not
enough to break PrivateSend privacy, it could still provide some additional
information for other methods like cluster analysis and help to narrow results.

With Random Round Mixing, an input will be mixed to N rounds like prior. However,

> at this point, SHA256(input, salt) will be calculated and roughly a half of inputs

Author: > it isn't clear from this description what role the SHA256 calculation plays in the process.

Would smth like this
```suggestion
at this point, a salted hash of each input is going to be used to pick roughly a half of inputs which
```
be better maybe?

> at this point, SHA256(input, salt) will be calculated and roughly a half of inputs
> will be mixed for one more round and this rule will be applied to newly created

Reviewer: ```suggestion
at this point, a salted hash of each input is used to pick ~50% of inputs to
be mixed for another round. This rule is then executed again on the new inputs.```

> inputs again and so on. This results in an exponential decay where if you mix a set

Reviewer: ```suggestion
This results in an exponential decay where if you mix a set
```

PoSe punishment/banning
-----------------------
If 80% of all LLMQ members voted for the same member to be bad, it is excluded
in the final stages of the DKG. This causes the bad masternode to get PoSe punished
and then eventually PoSe banned.

Network performance improvements
--------------------------------
This version of Dash Core includes multiple optimizations to the network and p2p message
handling code. The most important one is the introduction of `epoll` on linux-based
systems. This removes most of the CPU overhead caused by the sub-optimal use of `select`,
which could easily use up 50-80% of the CPU time spent in the network thread when many
connections were involved.

Since these optimizations are exclusive to linux, it is possible that masternodes hosted
on windows servers will be unable to handle the network load and should consider migrating
to a linux based operating system.

Other improvements were made to the p2p message handling code, so that LLMQ
related connections do less work than full/normal p2p connections.

Wallet files
------------
The `--wallet=<path>` option now accepts full paths instead of requiring

> wallets to be located in the -walletdir directory.

Reviewer 2: ```suggestion
wallets to be located in the `-walletdir` directory.
```

If `--wallet=<path>` is specified with a path that does not exist, it will now

> create a wallet directory at the specified location (containing a wallet.dat

Reviewer 2: Probably not absolutely needed but could make it a bit more clear together with the next comment where we have `database/log.?????????` instead of database/log.?????????? which imo looks weird in the middle of a text. 
```suggestion
create a wallet directory at the specified location (containing a `wallet.dat`
```

Or maybe just wrap all of them in "" ?

> data file, a db.log file, and database/log.?????????? files) instead of just

Reviewer 2: ```suggestion
data file, a `db.log` file, and `database/log.??????????` files) instead of just
```

creating a data file at the path and storing log files in the parent
directory. This should make backing up wallets more straightforward than
before because the specified wallet path can just be directly archived without
having to look in the parent directory for transaction log files.

For backwards compatibility, wallet paths that are names of existing data files
in the `--walletdir` directory will continue to be accepted and interpreted the
same as before.

When Dash Core is not started with any `--wallet=<path>` options, the name of
the default wallet returned by `getwalletinfo` and `listwallets` RPCs is
now the empty string `""` instead of `"wallet.dat"`. If Dash Core is started
with any `--wallet=<path>` options, there is no change in behavior, and the
name of any wallet is just its `<path>` string.

PrivateSend coin management improvements
----------------------------------------
A new algorithm for the creation of mixing denominations was implemented which
should reduce the number of the smallest inputs created and give users more
control on soft and hard caps. A much more accurate fee management algorithm was

> also implemented which should fix issues for users who have custom fees

Reviewer 2: ```suggestion
also implemented which fixes issues user had with custom fees
```

Reviewer: doesn't really make sense

you have to perform a reindex or re-sync the whole chain.

### Downgrade of masternodes to < 0.16

Starting with this release, masternodes will verify protocol versions of other
masternodes. This will cause PoSe punishment/banning of downgraded masternodes,
so it is not recommended to downgrade masternodes.

Notable changes
===============

Block Reward Reallocation
-------------------------
This version implements Block Reward Reallocation which was proposed in order
to slow the growth rate of Dash’s circulating supply by encouraging the
formation of masternodes and was voted in by the network. The resulting allocation
will split all non-proposal block rewards 40% toward miners and 60% toward
masternodes in the end-state once the transition period is complete.

The reallocation will take place gradually over 4.5 years to avoid market volatility
and transition toward the new end-state allocation with minimal network disruption.
The proposed changes will occur every three superblock cycles until the reallocation
is complete, or approximately once per quarter.

Note that this is a hardfork which must be activated by miners. To do this they

> should start creating blocks signalling bit 5 in `version` field of the block header.

Reviewer 3: This is described at the end of the following section

implements BIP9-like dynamic activation thresholds which drop from some initial
level to a minimally acceptable one over time at an increasing rate providing
a safe non-blocking way of activating proposals.

This mechanism applies to the Block Reward Reallocation proposal mentioned above.

Concentrated Recovery
---------------------
In the current system, signature shares are propagated to all LLMQ members,
until one of them has enough shares collected to recover the signature. Until
this recovered signature is propagated in the LLMQ, all members will keep
propagating shares and verifying each one. This causes quite some load on the
LLMQ, which will be avoided with the new system.

The new system sends all shares to only a single deterministically selected node,
so that this node can recover the signature and propagate the recovered signature.
This way only the recovered signature needs to be propagated and verified by all
members. After sending the share to this single node, each member waits for some
time and repeats sending it to another deterministically selected member. This
process is repeated until a recovered signature is finally created and propagated.

> The new system is activated with the newly added `SPORK_21_QUORUM_ALL_CONNECTED`.

Reviewer: I feel like we should justify the addition of a new spork/why this feature is tied to a spork

Increased number of masternode connections
------------------------------------------
To implement "Concentrated Recovery", it is now required that all members of a LLMQ

> connect to all other members of the same LLMQ. This increases general connection count
> for masternodes a lot. These intra-quorum connections are less resource consuming than

Reviewer: ```suggestion
connect to all other members of the same LLMQ. This significantly increases general connection count
for masternodes. These intra-quorum connections are less resource consuming than
```

normal p2p connections as they only exchange LLMQ/masternode related messages, but
the hardware and network requirements will still be higher than before.

> This change will at first only be activated for the smaller LLMQs (50 members) and
> then later for the larger ones (400 members).

Reviewer: ```suggestion
This change will at first only be activated for the smaller LLMQs (50 members) and
then may later be activated for larger quorums (400 members).
```

> This is also controlled via `SPORK_21_QUORUM_ALL_CONNECTED`.

Reviewer: Should talk about the duality of the spork here? difference between 0 being all quorums 1 being small quorums?

for `MNAUTH` to succeed and then disconnecting again.

After this process, each member of a LLMQ knows which members are unable to accept
connections, after which they will vote on these members to be "bad".

Masternode Minimum Protocol Version Checks
------------------------------------------
Members of LLMQs will now also check all other members for minimum protocol versions
while in DKG. If a masternode determines that another LLMQ member has a protocol
version that is too low, it will vote for the other member to be "bad".

PoSe punishment/banning
-----------------------
If 80% of all LLMQ members voted for the same member to be bad, it is excluded
in the final stages of the DKG. This causes the bad masternode to get PoSe punished
and then eventually PoSe banned.

Network performance improvements
--------------------------------
This version of Dash Core includes multiple optimizations to the network and p2p message
handling code. The most important one is the introduction of `epoll` on linux-based
systems. This removes most of the CPU overhead caused by the sub-optimal use of `select`,
which could easily use up 50-80% of the CPU time spent in the network thread when many

> connections were involved.

Reviewer: ```suggestion
connections were involved.

Since these optimizations are exclusive to linux, it is possible that masternodes hosted on windows servers will be unable to handle the network load and should consider migrating to a linux based operating system. 
```

data file, a db.log file, and database/log.?????????? files) instead of just
creating a data file at the path and storing log files in the parent
directory. This should make backing up wallets more straightforward than
before because the specified wallet path can just be directly archived without
having to look in the parent directory for transaction log files.

For backwards compatibility, wallet paths that are names of existing data files
in the `--walletdir` directory will continue to be accepted and interpreted the
same as before.

When Dash Core is not started with any `--wallet=<path>` options, the name of
the default wallet returned by `getwalletinfo` and `listwallets` RPCs is
now the empty string `""` instead of `"wallet.dat"`. If Dash Core is started
with any `--wallet=<path>` options, there is no change in behavior, and the
name of any wallet is just its `<path>` string.

PrivateSend coin management improvements
----------------------------------------
A new algorithm for creation of mixing denominations was implemented which
should reduce the number of the smallest inputs created and give users more

> control on soft and hard caps. A much more accurate fee management was

Reviewer: mention cmd options for these caps?

> control on soft and hard caps. A much more accurate fee management was

Reviewer: ```suggestion
control on soft and hard caps. A much more accurate fee management algortihm was
```

also implemented which should fix issues for users who have custom fees

> specified in wallet config or in times when network is under a stress.

Reviewer: ```suggestion
specified in wallet config or in times when network is under load.
```

> There is a new tab in GUI called "PrivateSend" which allows spending fully

Reviewer: ```suggestion
There is a new "PrivateSend" tab in the GUI which allows spending fully
```

mixed coins only. CoinControl feature was also improved to display coins

> based on the tab it was opened at, to protect users from accidentally spending

Reviewer: maybe?
```suggestion
based on the tab it was opened in, to protect users from accidentally spending
```

> mixed and non-mixed coins in one transaction and to give better overview of

Reviewer: ```suggestion
mixed and non-mixed coins in the same transaction and to give better overview of
```

spendable mixed coins.

PrivateSend Random Round Mixing
-------------------------------
Some potential attacks on PrivateSend assume that all inputs had been mixed
for the same number of rounds (up to 16). While this assumtion alone is not
enough to break PrivateSend privacy, it could still provide some additional
information for other methods like cluster analysis and help to narrow results.

With Random Round Mixing, an input will be mixed to N rounds like prior. However,
at this point, SHA256(input, salt) will be calculated and roughly a half of inputs
will be mixed for one more round and this rule will be applied to newly created

> inputs again and so on. This results in an exponential decay where if you mix a set

Reviewer: @thephez I feel like this could be worded better, but I'm not sure how

Reviewer 3: I don't think it's too bad, but it isn't clear from this description what role the SHA256 calculation plays in the process.

given the privacy gains.

GUI changes
-----------
Many dialogs, tabs, icons, colors and various interface elements were reworked
to improve user experience and to make wallet look more consistent. The wallet
GUI is also much more flexible now. For regular users there is a new "Appearance"
dialog and a corresponding tab in Options which allows them to pick a theme and
to tweak various font parameters. This feature specifically should help users
who had issues on some systems previously.

For advanced users and developers there is a new way to control wallet look
by specifying a path to css files via `--custom-css-dir`. Adding `--debug-ui`
will force it to reload these css files and to apply all changes live.

The "Pay To" field on "Send" and "PrivateSend" tabs accepts not only plain
Dash addresses but also Dash URIs now. The Dash address and the amount from
the URI are assigned to corresponding fields automatically.

Sporks
------
Two new sporks were introduced in this version:
- `SPORK_21_QUORUM_ALL_CONNECTED` allows to control Concentrated Recovery and
Masternode Probing activation;
- `SPORK_22_PS_MORE_PARTICIPANTS` rises the number of users that can participate

> in a single PrivateSend mixng transaction to 20.

Reviewer: If I remember, 0.16 also changes mixing from being a random number of participants between 3 and 5 to being as many participants as are available with a minimum of 3 and a max of 20

> in a single PrivateSend mixng transaction to 20.

Reviewer: ```suggestion
in a single PrivateSend mixing transaction to 20.
```

Reviewer: Looks like this is out of date
